### PR TITLE
DCCLIP-350: Add Bitbucket Mesh to Bitbucket Helm chart

### PR DESCRIPTION
**Pull request by bianchi2**
# Add Bitbucket Mesh

This PR adds [Bitbucket Mesh](https://confluence.atlassian.com/bitbucketserver/bitbucket-mesh-1128304351.html) to Bitbucket Helm chart.

### What is added:

* Bitbucket Mesh StatefulSet (3 replicas, no scaling or auto-scaling)
* Service per StatefulSet pod (using label selector `statefulset.kubernetes.io/pod-name`)
* JVM ConfigMap
* post-install jobs to
  * register mesh nodes with the server
  * make mesh the default for all new Bitbucket repositories

### Defaults

* Mesh is off by default, and can be deployed by setting `bitbucket.mesh.enabled=true`
* volumeClaimTemplates are added by default, which means mesh home directory is persisted. Do not deploy Mesh without persisting data.
* k8s jobs to register mesh nodes and make mesh default for all new repos are off by default and must be explicitly enabled.

## Mesh Node Registration

Post-install jobs are off by default, and can only be created if a secret with Bitbucket sysadmin username is defined in values. Jobs are annotated with post-install hooks which means they are unavailable for users with installed Bitbucket releases. While jobs do check exit code of the REST calls they make, it's a CI mostly feature, and we recommend adding mesh nodes manually in the Bitbucket admin UI.

### Scaling and AutoScaling

Not supported. Currently, Bitbucket server can only work with a set of 3 mesh nodes. Fewer nodes result in mesh cluster degradation, while deploying more mesh pods does not have any effect, as extra mesh nodes do not receive the repositories. For the above reason, manual scaling or HPA aren't supported.

### Server-Mesh Comms

gRPC comms aren't encrypted out of the box. It is recommended to use internal URLs, i.e. service name based which are resolved only in the cluster internal network. It's possible to create Mesh services as LoadBalancer type and terminate TLS at loadbalancer, however, it won't give much benefit compared to comms using service names (still http from LB to mesh svc). Mesh hasn't been tested with Istio or similar mutual TLS solutions.

TLS can be enabled in Mesh itself by specifying paths to key and crt in env vars, however, terminating tls on mesh side can impact performance.

### Multi AZ Deployments

TopologySpreadConstraints, NodeSelector and Tolerations can be configured for Mesh StatefulSet, i.e. it is possible to deploy each of the mesh pods to a node in a different AZ. However, one should bear in mind that if a pod is evicted and there are no suitable nodes in the required AZ, the pod will be stuck in Pending, and quorum will be lost.

### Do Mesh Resources Depend on Bitbucket values?

Mesh resources can be configured with `bitbucket.mesh.*` values, i.e. they are independent. The only dependent resource is fluentd configmap which is shared with Bitbucket (mesh config is added if mesh is enabled).

## Checklist
- [x] I have added unit tests
- [x] I have applied the change to all applicable products
- [ ] I have added the change description to the `changelog.md` and `Chart.yaml` files
- [x] (Atlassian only) I have run the E2E test (if applicable)
- [ ] (Atlassian only) Internal Bamboo CI is passing

### CODE REVIEW

File: src/main/charts/bitbucket/values.yaml [hunk 1]
```diff
@@ -752,6 +752,178 @@ bitbucket:
...
+      #
+      repository: atlassian/bitbucket-mesh
+
+      # -- Image pull policy
+      #
+      pullPolicy: IfNotPresent
+
+      # -- The docker image tag to be used - defaults to the Chart appVersion
+      #
+      tag: "1.3.0"
```
Comment (bianchi2) [inline src/main/charts/bitbucket/values.yaml:791]: TODO: Once 1.5.0 is pushed to DockerHub, update the version. 1.3.0 is the latest available.

File: src/main/charts/bitbucket/templates/statefulset-mesh.yaml [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        {{- if .Values.fluentd.enabled }}
+        checksum/config-fluentd: {{ include (print $.Template.BasePath "/configmap-fluentd.yaml") . | sha256sum }}
+        {{- end }}
+        {{- include "bitbucket.mesh.podAnnotations" . | nindent 8 }}
+      labels:
+        app.kubernetes.io/name: {{ include "common.names.name" . }}-mesh
+        app.kubernetes.io/instance: {{ .Release.Name }}
+        {{- include "bitbucket.mesh.podLabels" . | nindent 8 }}
+    spec:
+      serviceAccountName: {{ include "bitbucket.serviceAccountName" . }}
```
Comment (bianchi2) [inline src/main/charts/bitbucket/templates/statefulset-mesh.yaml:33]: I think it's ok to run with the default serviceAccount here.

File: src/test/config/bitbucket/helm_parameters [hunk 3]
```diff
@@ -11,7 +11,7 @@ ELASTICSEARCH_APP_VERSION=7.16.3
 LOG_DOWNLOAD_DIR=${project.build.directory}/logs
 TARGET_NAMESPACE=${kubernetes.target.namespace}
 DOCKER_IMAGE_REGISTRY=${dockerImage.registry}
-DOCKER_IMAGE_VERSION=${dockerImage.version}
+DOCKER_IMAGE_VERSION=8.6.1
```
Comment (bianchi2) [inline src/test/config/bitbucket/helm_parameters:14]: Override default version to test mesh

File: src/test/config/bitbucket/values-EKS.yaml [hunk 4]
```diff
@@ -55,3 +59,7 @@ affinity:
 
 nodeSelector:
   node_purpose: ci
+
+image:
+  # test Bitbucket mesh
+  tag: 8.6.1
```
Comment (bianchi2) [inline src/test/config/bitbucket/values-EKS.yaml:65]: Override default version to test mesh

File: src/test/config/bitbucket/values-KITT.yaml [hunk 5]
```diff
@@ -25,3 +25,15 @@ volumes:
...
+  mesh:
+    podAnnotations:
+      "atlassian.com/business_unit": "server_engineering"
+    enabled: true
+    nodeAutoRegistration: true
+    setByDefault: true
+
+image:
+  # test Bitbucket mesh
+  tag: 8.6.1
```
Comment (bianchi2) [inline src/test/config/bitbucket/values-KITT.yaml:39]: Override default version to test mesh

File: src/test/scripts/helm_uninstall.sh [hunk 6]
```diff
@@ -43,5 +43,7 @@ if [ "$shouldCleanNfsPod" == true ]; then
   kubectl delete -n "$TARGET_NAMESPACE" pvc -l "app.kubernetes.io/instance=$PRODUCT_RELEASE_NAME-nfs" --ignore-not-found=true 2>/dev/null || true
 fi
 
+kubectl delete -n "$TARGET_NAMESPACE" jobs -l app.kubernetes.io/instance="${PRODUCT_RELEASE_NAME}" --ignore-not-found=true
```
Comment (bianchi2) [inline src/test/scripts/helm_uninstall.sh:46]: Leftover from experiments where jobs aren't removed after they are completed

File: src/main/charts/bitbucket/templates/job-mesh-add-nodes.yaml [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        {{- include "bitbucket.mesh.podAnnotations" . | nindent 8 }}
+      labels:
+        app.kubernetes.io/name: {{ include "common.names.name" . }}-mesh-job
+        app.kubernetes.io/instance: {{ .Release.Name }}
+        {{- include "bitbucket.mesh.podLabels" . | nindent 8 }}
+    spec:
+      containers:
+        - name: bitbucket-mesh-register-services
+          image: curlimages/curl
+          command: ["/scripts/register-mesh-node.sh"]
```
Comment (bianchi2) [inline src/main/charts/bitbucket/templates/job-mesh-add-nodes.yaml:28]: https://github.com/atlassian/data-center-helm-charts/pull/501/files#diff-68968aa1d23eb550c505be4c9bfac3f64becf9ac17e4e9ec0c7f204b32b087ec

Having commands in a configmap is better, especially if response code need to be checked.

File: src/test/config/bitbucket/helm_parameters (hunk 3 above)
Comment (nghazali) [reply src/test/config/bitbucket/helm_parameters:14]: Should we change this when the new version become the default one? In this case is good to document this somewhere to not forgotten when the default version changed in future.

File: docs/docs/examples/bitbucket/BITBUCKET_MESH.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
+# Bitbucket Mesh
+
+Bitbucket Mesh is a distributed, replicated, and horizontally scalable Git repository storage system, which increases performance and improves the resilience of Bitbucket.
+
+![bitbucket_mesh](../../assets/images/bitbucket_with_mesh.png)
+
+You can learn more details about Bitbucket Mesh on the [official documentation page](https://confluence.atlassian.com/bitbucketserver/bitbucket-mesh-1128304351.html){.external}.
```
Comment (mariasob) [inline docs/docs/examples/bitbucket/BITBUCKET_MESH.md:7]: Learn more details about BitBucket Mesh <- that can be the whole link here

File: docs/docs/examples/bitbucket/BITBUCKET_MESH.md [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+!!!tip "Recommendations for Mesh deployments"
+
+    **Bitbucket version**
+
+    By default the Helm charts target the latest Bitbucket LTS version. However
```
Comment (mariasob) [inline docs/docs/examples/bitbucket/BITBUCKET_MESH.md:13]: By default, the Helm charts target the latest Bitbucket LTS version. However,

File: docs/docs/examples/bitbucket/BITBUCKET_MESH.md [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    mesh is only supported from version 8.0. You will need to select an 8.x
```
Comment (mariasob) [inline docs/docs/examples/bitbucket/BITBUCKET_MESH.md:14]: Mesh is only supported from version 8.0. You will need to select an 8.x

File: docs/docs/examples/bitbucket/BITBUCKET_MESH.md [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    for available versions.
+
+    **Number of Mesh nodes**
+
+    In order for high-availability to be possible, we recommend having a minimum
+    of three Mesh nodes. There is no maximum on the number of nodes.
+
+    **Mesh node co-location**
+
+    Currently, we will not be supporting deploying Mesh nodes into multiple
```
Comment (mariasob) [inline docs/docs/examples/bitbucket/BITBUCKET_MESH.md:31]: We don't currently support deploying Mesh nodes into multiple

File: docs/docs/examples/bitbucket/BITBUCKET_MESH.md [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    availability zones. Just like the shared file system based deployments, the
+    Mesh nodes (that is, the repository storage) and the application nodes must
+    be co-located.
+
+    **Other Mesh deployment requirements**
+
+    For more details on the requirements and limitations of Mesh deployments see
```
Comment (mariasob) [inline docs/docs/examples/bitbucket/BITBUCKET_MESH.md:38]: For more details on the requirements and limitations of Mesh deployments, [check the Bitbucket Mesh FAQ](https://confluence.atlassian.com/enterprise/bitbucket-data-center-faq-776663707.html#BitbucketDataCenterFAQ-Mesh).

File: docs/docs/examples/bitbucket/BITBUCKET_MESH.md [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    [the Bitbucket Mesh FAQ](https://confluence.atlassian.com/enterprise/bitbucket-data-center-faq-776663707.html#BitbucketDataCenterFAQ-Mesh){.external}
+
+### Configuring your Bitbucket and Mesh deployment
+
+For backwards-compatibility the Helm charts default to Mesh being disabled. To
```
Comment (mariasob) [inline docs/docs/examples/bitbucket/BITBUCKET_MESH.md:43]: For backwards compatibility, the Helm charts default to Mesh being disabled. To

File: docs/docs/examples/bitbucket/BITBUCKET_MESH.md [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+enable it you will need to configure the service under `bitbucket:` stanza in
```
Comment (mariasob) [inline docs/docs/examples/bitbucket/BITBUCKET_MESH.md:44]: enable it, you'll need to ...(the rest stays the same)

File: docs/docs/examples/bitbucket/BITBUCKET_MESH.md [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+bitbucket:
+  mesh:
+    enabled: true
+    image:
+      version: <Mesh agent version>
+```
+
+### Adding the Mesh nodes to Bitbucket
+
+To enable the deployed Mesh nodes you need to add them to the Bitbucket Data
```
Comment (mariasob) [inline docs/docs/examples/bitbucket/BITBUCKET_MESH.md:61]: To enable the deployed Mesh nodes, add them to your Bitbucket Data

File: docs/docs/examples/bitbucket/BITBUCKET_MESH.md [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+Center instance via the administration interface. To do this you will need the
+service URL of each node; these are usually of the form `bitbucket-mesh-<num>`. Check the Kubernetes
+official documentation on how to get [DNS record for a
```
Comment (mariasob) [inline docs/docs/examples/bitbucket/BITBUCKET_MESH.md:64]: official documentation to learn how to get a [DNS record for a

File: docs/docs/examples/bitbucket/BITBUCKET_MESH.md [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+service](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#services){.external}.
+
+To connect the mesh node:
```
Comment (mariasob) [inline docs/docs/examples/bitbucket/BITBUCKET_MESH.md:67]: To connect the Mesh node:

File: docs/docs/examples/bitbucket/BITBUCKET_MESH.md [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+
+1. In your Bitbucket Data Center instance, navigate to **Administration** > **Git** > **Bitbucket Mesh**.
+1. Enter the URL of the Mesh node in the Node URL field (e.g. `http://bitbucket-mesh-1:7777`).
+1. (Optional) Enter a name for the Mesh node in the Node name field.
+1. Select Add Mesh node.
```
Comment (mariasob) [inline docs/docs/examples/bitbucket/BITBUCKET_MESH.md:72]: 1. Select **Add Mesh node**.

File: docs/docs/examples/bitbucket/BITBUCKET_MESH.md [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+
+For full details see [the documentation for configuring Mesh](https://confluence.atlassian.com/bitbucketserver/set-up-and-configure-mesh-nodes-1128304356.html#SetupandconfigureMeshnodes-ConnectingtheMeshnodetoBitbucket).
```
Comment (mariasob) [inline docs/docs/examples/bitbucket/BITBUCKET_MESH.md:74]: Learn more details about Mesh configuration <- this is the link

File: docs/docs/examples/bitbucket/BITBUCKET_MESH.md [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+
+### Migrating existing repositories to Mesh
+
+For migrating repositories to Mesh see [the official documentation](https://confluence.atlassian.com/bitbucketserver/migrate-repositories-to-bitbucket-mesh-1128304358.html).
```
Comment (mariasob) [inline docs/docs/examples/bitbucket/BITBUCKET_MESH.md:78]: Learn how to migrate repositories to Mesh <- this is the link

**Review by mariasob — CHANGES_REQUESTED** [review]
I've left some comments.

**Review by nanux — APPROVED** [review]

**Review by tarka — DISMISSED** [review]

**Review by tarka — APPROVED** [review]

**Pull request merged**